src/engine/BMInterfaceTournament.php

```diff
@@ -0,0 +1,987 @@
+                            ':button_id' => $buttonIdArray[0],
+                            ':n_players' => $nPlayers);
+        return self::$db->update_and_report_if_changed($query, $parameters);
+    }
+
+    /**
+     * Attempt to remove a player from a tournament
+     *
+     * @param int $userId
+     * @param int $tournId
+     * @return bool
+     */
+    protected function leave_tournament($userId, $tournId) {
+        try {
+            $tournament = $this->load_tournament($tournId);
+
+            if (!$this->validate_leave_tournament($userId, $tournId, $tournament)) {
+                return NULL;
+            }
+
+            $success = $this->remove_user_from_tournament($userId, $tournId, $tournament->nPlayers);
+
+            if (!$success) {
+                // something's gone wrong between validation and attempting to remove the user
+                $tournament = $this->load_tournament($tournId);
+                $validation = $this->validate_leave_tournament($userId, $tournId, $tournament);
+
+                if ($validation) {
+                    error_log('Paradoxical validation success in BMInterface::leave_tournament');
+                    $this->set_message('Tournament leave failed even though it should have succeeded');
+                }
+                return NULL;
+            }
+
+            $this->unwatch_tournament($userId, $tournId);
+        } catch (BMExceptionDatabase $e) {
+            $this->set_message('Cannot leave tournament because a player or tourn ID was not valid');
+            return NULL;
+        } catch (Exception $e) {
+            error_log(
+                'Caught exception in BMInterface::leave_tournament: ' .
+                $e->getMessage()
+            );
+            $this->set_message("Tournament leave failed: $e");
+            return NULL;
+        }
+
+        $this->update_tournament($tournId);
+        $this->set_message('You have successfully left this tournament');
+        return TRUE;
+    }
+
+    /**
+     * Validate whether a player can leave a tournament
+     *
+     * @param int $userId
+     * @param int $tournId
+     * @param BMTournament $tournament
+     * @return bool
+     */
+    protected function validate_leave_tournament($userId, $tournId, $tournament) {
+        if (is_null($tournament)) {
+            $this->set_message("Tournament $tournId does not exist");
+            return NULL;
+        }
+
+        if ($tournament->tournamentState > BMTournamentState::JOIN_TOURNAMENT) {
+            $this->set_message('The tournament has already started.');
+            return NULL;
+        }
+
+        if (!in_array($userId, $tournament->playerIdArray)) {
+            $this->set_message('You are not part of this tournament.');
+            return NULL;
+        }
+
+        if (count($tournament->playerIdArray) == $tournament->nPlayers) {
+            $this->set_message('You cannot leave when the tournament is full.');
+            return NULL;
+        }
```

Comment:
I'm just curious: is it a valid state of the world for a tournament to be full and still be in JOIN_TOURNAMENT state (outside of the DB transaction which filled the tournament, which is presumably not this transaction), or is that secretly an internal error (in which case we should at least log it)?

Reply:
Converted to error_log and thrown exception.